appcelerator/titanium_mobile
# [TIMOB-25941] Android: Optimized HTTPClient download performance

**JIRA:**
https://jira.appcelerator.org/browse/TIMOB-25941

**Summary:**
- Download performance improved by about 25%. (Depends on network speed/quality.)
- Increased response buffer size from 4kb to 8kb.
- Removed undocumented event property `blob` from callback `ondatastream()`.
  * Improves performance by avoiding extra buffer copy.
  * This undocumented property does not exist on iOS.
- Modified `responseText` property handling only create its string when accessed.
  * Was creating and transcoding the string for every response segment received. Was a huge performance hit.
  * Fixed `responseText` to provide total text downloaded like iOS instead of the buffered response segment's text.
- Synchronized file write to storage upon download completion.
  * Avoids race condition where reading `responseText` or `responseData` in `onload()` callback may provide an incomplete response.

**Test:**
@lokeshchdhry, I recommend that you test this at home. The office network is too congested and download speed varies wildly.
1. Acquire an Android device with Internet access.
2. Build and run the below code with Titanium 7.3.0.
3. Tap the "Download" button. 
4. Wait for the download to complete.
5. Note down the download duration time. (You may want to do this a few times and record the average.)
6. Build and run with this PR and repeat steps 3-5.
7. The download time should have improved.

```javascript
var url = "https://github.com/appcelerator-modules/ti.admob/releases/download/ios-2.2.0/ti.admob-iphone-2.2.0.zip";
var downloadFilePath = Ti.Filesystem.applicationDataDirectory + "downloaded.zip";

var window = Ti.UI.createWindow();
var button = Ti.UI.createButton({ title: "Download" });
var progressBar = Ti.UI.createProgressBar({
	min: 0,
	max: 100,
	bottom: "40dp",
	left: "5dp",
	right: "5dp",
});
button.addEventListener("click", function(e) {
	var startTime = new Date();
	var httpClient = Ti.Network.createHTTPClient({
		ondatastream: function(e) {
			progressBar.value = e.progress * 100;
		},
		onload: function(e) {
			var duration =  (new Date()) - startTime;
			Ti.API.info("@@@ onload() duration: " + duration + "ms");
			alert("Download Time: " + duration + "ms");
			progressBar.value = 100;
			button.enabled = true;
		},
		onerror: function(e) {
			Ti.API.info("@@@ onerror()");
			alert("Download Failed");
			button.enabled = true;
		},
	});
	progressBar.value = 0;
	button.enabled = false;
	httpClient.open("GET", url);
	httpClient.setRequestHeader("Accept-Encoding", "identity");
	httpClient.file = downloadFilePath;
	httpClient.send();
});
window.add(button);
window.add(progressBar);
window.open();
```
